asciidoc/modules/con_mg_changes-in-vertx-grpc.adoc

```diff
@@ -5,42 +5,123 @@ The following section describes the changes in {VertX} gRPC.
-              <artifactId>vertx-grpc-protoc-plugin</artifactId>
-              <version>${vertx.version}</version>
-              <mainClass>io.vertx.grpc.protoc.plugin.VertxGrpcGenerator</mainClass>
-            </protocPlugin>
-          </protocPlugins>
-        </configuration>
-      </plugin>
+    <!-- Vert.x 4.0 -->
+<plugin>
+  <groupId>org.xolstice.maven.plugins</groupId>
+  <artifactId>protobuf-maven-plugin</artifactId>
+  <configuration>
+    <protocArtifact>com.google.protobuf:protoc:3.2.0:exe:${os.detected.classifier}</protocArtifact>
+    <pluginId>grpc-java</pluginId>
+    <!-- NOTE: the gav coordinates point to the official compiler -->
+    <pluginArtifact>io.grpc:protoc-gen-grpc-java:${vertx.grpc.version}:exe:${os.detected.classifier}</pluginArtifact>
+    <protocPlugins>
+      <!-- NEW: a plugin is added to generate the Vert.x specific code -->
+      <protocPlugin>
+        <id>vertx-grpc-protoc-plugin</id>
+        <groupId>io.vertx</groupId>
+        <artifactId>vertx-grpc-protoc-plugin</artifactId>
+        <version>${vertx.version}</version>
+        <mainClass>io.vertx.grpc.protoc.plugin.VertxGrpcGenerator</mainClass>
+      </protocPlugin>
+    </protocPlugins>
+  </configuration>
+  ...
+</plugin>
+----
+
+== Migrating the generated code
+
+In {VertX} {v4}, the new compiler is used. When the new gRPC plugin is used, the generated code is not written in the same source file. This is because the compiler does not allow custom code generation on its base class. The plugins must generate a new class with a different name to save the code.
+
+In earlier releases of {VertX}, the older gRPC plugin would write the generated code in the same source file.
+
+For example, if you have the following descriptor:
+
+[source,proto]
+----
+service Greeter {
+  rpc SayHello (HelloRequest) returns (HelloReply) {}
+}
+----
+
+In {VertX} {v3x}, the code would be generated in the `GreeterGrpc` class.
+
+[source,java]
+----
+// 3.x
+GreeterGrpc.GreeterVertxImplBase service =
+  new GreeterGrpc.GreeterVertxImplBase() {
+      ...
+  }
+----
+
+In {VertX} {v4}, the code is generated in the `VertxGreeterGrpc` class.
+
+[source,java]
+----
+// 4.x
+VertxGreeterGrpc.GreeterVertxImplBase service =
+  new VertxGreeterGrpc.GreeterVertxImplBase() {
+      ...
+  }
+----
+
+== gRPC APIs support futures
+
+In {VertX} {v4}, the gRPC APIs support futures.  Using these APIs the gRPC plugin generates promisified applications. This makes it easier to reason about inputs and outputs.XXX
+
+The following example shows you XXX in {VertX} {v3x}.
+
+[source,java]
+----
+// 3.x
+GreeterGrpc.GreeterVertxImplBase service =
+  new GreeterGrpc.GreeterVertxImplBase() {
+        @Override
+        public void sayHello(HelloRequest request, Promise<HelloReply> future) {
+          future.complete(
+              HelloReply.newBuilder().setMessage(request.getName()).build());
+        }
+  }
+----
+
+The following example shows you XXX in {VertX} {v4}.
+
```

Comment:
@pmlopes Can you please help with text here?

Reply:
While in {VertX} {v3x} the greet response would be writen to an argument of the function, in {VertX} {v4} the asynchronous response is the method return value. Which reads simpler and the code flow is easy to follow.

Reply:
As discussed @pmlopes used the following sentence:
The following example shows use of futures in {VertX} {v4}.